File: crypto-curriculum/src/crypto/hmac.ts
/**
 * 挑战 2 —— HMAC-SHA-256（挑战 2 参考实现）。
 *
 * `crypto.createHmac('sha256', key)` 实现的是 RFC 2104 中的 HMAC。
 * 对于 SHA-256，标签长度固定为 32 字节。我们同时暴露 `sign` 与 `verify`，
 * 其中 `verify` 使用 `crypto.timingSafeEqual` 进行常数时间比较。
 *
 * 为什么要常数时间：在标签上使用非常数时间的 `==` 会通过时延泄漏首个
 * 差异字节的位置。远程攻击者已经在实际中通过这种方式逐字节地恢复
 * 了面向网络的 HMAC 标签。
 */

import { createHmac, timingSafeEqual } from 'node:crypto';
import type { Mac } from './contracts.js';

export const HmacSha256: Mac = {
  tagLength: 32,

  sign(key, message) {
    const mac = createHmac('sha256', Buffer.from(key));
    mac.update(Buffer.from(message));
    return new Uint8Array(mac.digest());
  },

  verify(key, message, tag) {
    const expected = this.sign(key, message);
    if (expected.length !== tag.length) return false;
    return timingSafeEqual(Buffer.from(expected), Buffer.from(tag));
  },
};

File: crypto-curriculum/modules/11-rust/src/hmac.rs
//! HMAC-SHA-256 wrapper around `hmac::Hmac<sha2::Sha256>`.

use hmac::{Hmac, Mac};
use sha2::Sha256;

pub struct HmacSha256;

impl HmacSha256 {
    pub const TAG_LEN: usize = 32;

    pub fn sign(key: &[u8], message: &[u8]) -> [u8; 32] {
        let mut mac = <Hmac<Sha256>>::new_from_slice(key)
            .expect("HMAC accepts keys of any length");
        mac.update(message);
        mac.finalize().into_bytes().into()
    }

    pub fn verify(key: &[u8], message: &[u8], tag: &[u8]) -> bool {
        let expected = Self::sign(key, message);
        // Constant-time compare via byte-wise XOR-fold.
        if expected.len() != tag.len() {
            return false;
        }
        let mut diff = 0u8;
        for i in 0..expected.len() {
            diff |= expected[i] ^ tag[i];
        }
        diff == 0
    }
}

#[cfg(test)]
mod tests {
    use super::HmacSha256;

    #[test]
    fn round_trip() {
        let k = [0xab; 32];
        let m = b"the message";
        let t = HmacSha256::sign(&k, m);
        assert!(HmacSha256::verify(&k, m, &t));
        assert!(!HmacSha256::verify(&k, m, &{
            let mut t = t;
            t[7] ^= 0x01;
            t
        }));
    }
}
